dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -125,21 +125,22 @@ LimitedChannel getCurrentChannel() {
 
     @Override
     public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
-        return executeInternal(endpoint, request, 1);
+        return executeInternal(endpoint, request, currentHost.get(), 1);
     }
 
-    private Optional<ListenableFuture<Response>> executeInternal(Endpoint endpoint, Request request, int depth) {
-        int currentIndex = currentHost.get();
+    private Optional<ListenableFuture<Response>> executeInternal(
+            Endpoint endpoint, Request request, int currentIndex, int depth) {
         LimitedChannel channel = nodeList.get(currentIndex);
 
         Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        // Simply reroute the request to the fist available channel rather than waiting for the pinned channel to
+        // become available. We attempt to route the request at most nodeList.size() times, but due to reshuffling we
+        // are not guaranteed to attempt every distinct channel.
```

Comment:
In this case shouldn’t we fail and let the retryer do it’s thing?

Reply:
Further simplifying the logic and simply returning the pinned channels response is totally an option. I find it difficult to determine which approach would be better, since doing so would require making certain assumptions about the retry behaviour and encode a trade off between locality and latency which may not be universally true.

Reply:
As written I’m worried the proposed chang would further complicate the algorithm because it differentiates changing the pinned channel from attempting another channel without changing the pin. This could result in undesirable behavior for subsequent calls.
I think we need to decide whether or not a limited result is considered a failure in the context a pin until error. If so, the existing implementation is reasonable, but if not we should return the limited (but not failed) result. In the future we may implement a blocking limit to avoid burning retries unnecessarily and failing when retries aren’t allowed.
Pin until error with overflow is difficult for me to reason about, I’m not sure what the benefits are over rotating the pinned channel.